services/repository/files/update.go

```diff
@@ -481,181 +481,177 @@ func CreateOrUpdateFile(ctx context.Context, t *TemporaryUploadRepository, file
-			return err
-		}
-		exist, err := contentStore.Exists(lfsMetaObject.Pointer)
-		if err != nil {
-			return err
-		}
-		if !exist {
-			var lfsContentReader io.Reader
-			if file.Operation != "rename" {
-				if _, err := file.ContentReader.Seek(0, io.SeekStart); err != nil {
-					return err
-				}
-				lfsContentReader = file.ContentReader
-			} else {
-				if lfsContentReader, err = oldEntry.Blob().DataAsync(); err != nil {
-					return err
-				}
-				defer lfsContentReader.(io.ReadCloser).Close()
-			}
+	if writeObjectRet.LfsContent == nil {
+		return nil // No LFS pointer, so nothing to do
+	}
+	defer writeObjectRet.LfsContent.Close()
 
-			if err := contentStore.Put(lfsMetaObject.Pointer, lfsContentReader); err != nil {
-				if _, err2 := git_model.RemoveLFSMetaObjectByOid(ctx, repoID, lfsMetaObject.Oid); err2 != nil {
-					return fmt.Errorf("unable to remove failed inserted LFS object %s: %v (Prev Error: %w)", lfsMetaObject.Oid, err2, err)
-				}
-				return err
-			}
+	// Now we must store the content into an LFS object
+	lfsMetaObject, err := git_model.NewLFSMetaObject(ctx, repoID, writeObjectRet.LfsPointer)
+	if err != nil {
+		return err
+	}
+	if exist, err := contentStore.Exists(lfsMetaObject.Pointer); err != nil {
+		return err
+	} else if exist {
+		return nil
+	}
+
+	err = contentStore.Put(lfsMetaObject.Pointer, writeObjectRet.LfsContent)
+	if err != nil {
+		if _, errRemove := git_model.RemoveLFSMetaObjectByOid(ctx, repoID, lfsMetaObject.Oid); errRemove != nil {
+			return fmt.Errorf("unable to remove failed inserted LFS object %s: %v (Prev Error: %w)", lfsMetaObject.Oid, errRemove, err)
 		}
 	}
+	return err
+}
 
-	return nil
+func checkIsLfsFileInGitAttributes(ctx context.Context, t *TemporaryUploadRepository, paths []string) (ret []bool, err error) {
+	attributesMap, err := attribute.CheckAttributes(ctx, t.gitRepo, "" /* use temp repo's working dir */, attribute.CheckAttributeOpts{
+		Attributes: []string{attribute.Filter},
+		Filenames:  paths,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range paths {
+		isLFSFile := attributesMap[p] != nil && attributesMap[p].Get(attribute.Filter).ToString().Value() == "lfs"
+		ret = append(ret, isLFSFile)
+	}
+	return ret, nil
 }
 
-func createOrUpdateFileHash(ctx context.Context, t *TemporaryUploadRepository, file *ChangeRepoFile, hasOldBranch bool) (string, *lfs.Pointer, error) {
+type writeRepoObjectRet struct {
+	ObjectHash string
+	LfsContent io.ReadCloser // if not nil, then the caller should store its content in LfsPointer, then close it
+	LfsPointer lfs.Pointer
+}
+
+// writeRepoObjectForCreateOrUpdate hashes the git object for create or update operations
+func writeRepoObjectForCreateOrUpdate(ctx context.Context, t *TemporaryUploadRepository, file *ChangeRepoFile) (ret *writeRepoObjectRet, err error) {
+	ret = &writeRepoObjectRet{}
 	treeObjectContentReader := file.ContentReader
-	var lfsPointer *lfs.Pointer
-	if setting.LFS.StartServer && hasOldBranch {
```

Comment:
Another question: should we really check `hasOldBranch`? (I removed it in this PR)